src/shared/mount-util.c

```diff
@@ -475,67 +475,86 @@ int bind_remount_one_with_mountinfo(
         return 0;
 }
 
-int mount_move_root(const char *path) {
-        assert(path);
-
-        if (chdir(path) < 0)
-                return -errno;
-
-        if (mount(path, "/", NULL, MS_MOVE, NULL) < 0)
-                return -errno;
-
-        if (chroot(".") < 0)
-                return -errno;
-
-        return RET_NERRNO(chdir("/"));
-}
-
-int mount_pivot_root(const char *path) {
-        _cleanup_close_ int fd_oldroot = -EBADF, fd_newroot = -EBADF;
+int mount_switch_root(const char *path, MountAttrPropagationType type) {
+        _cleanup_close_ int fd_newroot = -EBADF;
+        unsigned int flags;
+        const char *prop_name;
 
         assert(path);
 
-        /* pivot_root() isn't currently supported in the initramfs. */
-        if (in_initrd())
-                return mount_move_root(path);
-
-        fd_oldroot = open("/", O_PATH|O_DIRECTORY|O_CLOEXEC|O_NOFOLLOW);
-        if (fd_oldroot < 0)
-                return log_debug_errno(errno, "Failed to open old rootfs");
+        switch (type) {
+        case PROPAGATION_INHERIT:
+                flags = 0;
+                prop_name = "inherited";
+                break;
+        case PROPAGATION_PRIVATE:
+                flags = MS_PRIVATE;
+                prop_name = "private";
+                break;
+        case PROPAGATION_DEPENDENT:
+                flags = MS_SLAVE;
+                prop_name = "dependent";
+                break;
+        case PROPAGATION_SHARED:
+                flags = MS_SHARED;
+                prop_name = "shared";
+                break;
+        default:
+                assert_not_reached();
+        }
 
-        fd_newroot = open(path, O_PATH|O_DIRECTORY|O_CLOEXEC|O_NOFOLLOW);
+        fd_newroot = open(path, O_PATH | O_DIRECTORY | O_CLOEXEC | O_NOFOLLOW);
         if (fd_newroot < 0)
                 return log_debug_errno(errno, "Failed to open new rootfs '%s': %m", path);
 
         /* Change into the new rootfs. */
         if (fchdir(fd_newroot) < 0)
                 return log_debug_errno(errno, "Failed to change into new rootfs '%s': %m", path);
 
-        /* Let the kernel tuck the new root under the old one. */
-        if (pivot_root(".", ".") < 0)
-                return log_debug_errno(errno, "Failed to pivot root to new rootfs '%s': %m", path);
+        if (in_initrd()) {
```

Comment:
so, this is an interesting change. The thing is we allow "second level initrds". i.e. minimal initrds which transition into larger initrds which then transition into the host. from the kernel persepctive only the first one really is an initrd. the secondary ones are not. However /etc/initrd-release is a concept that can be (and should be) used on both primary and secondary initrds.

Not sure what a better test would be here. One option might be the dumb approach, i.e. to try pivot_root() and if that fails with EPERM fall back to MS_MOVE?

Or can we somehow detect actual initrds in a better way, i.e. something in stat() or statfs() that reveals this is the real initrd, and not 2nd level one?

(btw, a certain project at MSFT uses second level initrds, if i remember correctly)

Reply:
Afaik, there's only one sure way to detect an initramfs:

```c
bool detect_ramfs_rootfs(void)                                                                      
{                                                                                                                                                                                                                    
        __do_free char *line = NULL;                                                                
        __do_free void *fopen_cache = NULL;                                                         
        __do_fclose FILE *f = NULL;                                                                 
        size_t len = 0;                                                                             
                                                                                                    
        f = fopen_cached("/proc/self/mountinfo", "re", &fopen_cache);                               
        if (!f)                                                                                     
                return false;                                                                       
                                                                                                    
        while (getline(&line, &len, f) != -1) {                                                     
                int i;                                                                              
                char *p, *p2;                                                                       
                                                                                                    
                for (p = line, i = 0; p && i < 4; i++)                                              
                        p = strchr(p + 1, ' ');                                                     
                if (!p)                                                                             
                        continue;                                                                   
                                                                                                    
                p2 = strchr(p + 1, ' ');                                                            
                if (!p2)                                                                            
                        continue;                                                                   
                *p2 = '\0';                                                                         
                if (strequal(p + 1, "/")) {                                                         
                        /* This is '/'. Is it the ramfs? */                                         
                        p = strchr(p2 + 1, '-');                                                    
                        if (p && strnequal(p, "- rootfs ", 9))                                      
                                return true;                                                        
                }                                                                                   
        }                                                                                           
                                                                                                    
        return false;                                                                               
}
```